credentials/tls/certprovider/provider.go

```diff
@@ -87,18 +87,43 @@ type StableConfig interface {
 // the latest secrets, and free to share any state between different
 // instantiations as they deem fit.
 type Provider interface {
-	// KeyMaterial returns the key material sourced by the provider.
-	// Callers are expected to use the returned value as read-only.
-	KeyMaterial(ctx context.Context) (*KeyMaterial, error)
+	// KeyMaterialReader returns a reader to read key material sourced by the
+	// provider. Users should call the Close() method on the returned reader
+	// when they are no longer interested in the underlying key material.
+	KeyMaterialReader(opts KeyMaterialOptions) (KeyMaterialReader, error)
 
 	// Close cleans up resources allocated by the provider.
 	Close()
 }
 
+// KeyMaterialReader is used to read key material at handshake time.
+// Implementations must be thread-safe.
+type KeyMaterialReader interface {
```

Comment:
What do you think: should we simplify things and just delete this interface and require people to use `Distributor`s directly?

Reply:
I agree that it is one extra interface as it stands now. But it does have a few benefits:
* We are only exposing the read-side of the `Distributor` which is all that the users of this interface need.
* Defining the interface also keeps the option of alternate implementations open, where `Provider`s don't want to use our `Distributor`.
* Conceptually very similar to many other `Reader` interfaces that we are used to.

I guess when we started, we had two contrasting proposals:
* One where we had the `Source` and the `Sink` which represented the two sides of the `Distributor` and
* One where the `Distributor` was completely optional, and users dealt directly with the `Provider`.

I feel the current approach is a middle ground between the two.

Reply:
I don't see things simplifying by much, if we combine the `Builder` and the `Provider`. Maybe I'm missing something?

1. On the Producer side, whether they implement two interfaces (`Builder` and `Provider`), or they implement a combined interface `ProviderBuilder` (for lack of a better name), they need to implement almost the same functionality (which is actually building a provider, parsing config, and the business logic of the provider).

 2. On the consumer side, gRPC would still have to work with the same three APIs that you've mentioned.

What combining the `Builder` and the `Provider` does it that it moves some of the sharing functionality currently provided by the `Store` into the `ProviderBuilder`. The `Store` now handles sharing at both the `Provider` level and the `KeyMaterialReader` level. 

And regarding `1.b`, the `Provider` could chose to override the `Close()` method on the `Distributor`, thereby avoiding one goroutine per `Distributor` to notice closure. But I'm not sure it is worth it, because in most cases, we are not going to have very many certificate names.

Reply:
I'm not sure the extra functionality of the store caching an intermediate type is worth the extra complexity that comes with having another "primary" symbol in this package.  The `Builder` can cache its own intermediate data as needed.  The names were pretty intuitive before: "a certprovider.Provider provides certs".  Now there's another level of indirection that makes it harder to follow: "a certprovider.Provider provides a KeyMaterialReader which provides certs".

This design also operates under the assumption that you will want one object per config and that object is capable of handling many names.  But it's possible that you need one object for every config+name combo.  It's also possible that you can share one object among all of the configs.  Yes, this design can handle those use cases, but it brings unnecessary complexity for them to implement all the types the way this design requires them.

Reply:
So, I'm a little confused here :)
Do you want to combine the `Builder` and the `Provider` or do you want to get rid of the extra `KeyMaterialReader` type, or both?

Reply:
> Do we really need a separate type for `Builder` & `Provider`? Or can we combine them? If we can combine them, maybe `Provider` can go back to exactly what it was before, and that would simplify things. I.e. `Build` accepts a third parameter for the cert name, and the `Builder` is responsible for sharing something it owns when appropriate.

So, if we change the `Build()` method to also take a `KeyMaterialOptions`, it would look as follows:
`Build(config, opts) Provider`.

The issues I see here are as follows:
1. `Builder` would have to wrap the `Provider` that it returns so that it can implement reference counting. 
2. How would we connect the `certName` to the returned `Provider`, which now simply has a `KeyMaterial()` method? The `Builder` and the `Provider` can arrange for that through APIs that are not exported. But this puts a lot of burden on every plugin implementor to get things right.
3. The `Close` method on the returned `Provider` is ambiguous. Does it mean to close the whole `Provider` or simply the close the internals of the `Provider` related to the `certName` specified in the call to `Build`?

Reply:
> Builder would have to wrap the Provider that it returns so that it can implement reference counting.

I was thinking the store would still cache and refcount `Provider`s, just as it does `KeyMaterialReader`s today.

> How would we connect the certName to the returned Provider, which now simply has a KeyMaterial() method? The Builder and the Provider can arrange for that through APIs that are not exported. But this puts a lot of burden on every plugin implementor to get things right.

I don't understand your concerns here.  `Build` takes a `certName` and returns a `Provider`.  That `Provider` would know what `certName` it's responsible for, as the `Builder` would initialize it with that data.

> The Close method on the returned Provider is ambiguous. Does it mean to close the whole Provider or simply the close the internals of the Provider related to the certName specified in the call to Build?

`Close` would close that `Provider` instance.  That `Provider` only provides key material for one `certName`.  If the `Builder` wishes to share components (e.g. a gRPC client) among `Provider`s, it will need to refcount those and close them when the final `Provider` referencing them is closed.

Reply:
> Do you want to combine the Builder and the Provider or do you want to get rid of the extra KeyMaterialReader type, or both?

I am suggesting the only change from the previous design is that `certName` is passed to `Builder.Build`, and the `certName` is also used by the store to be able to find a cached `Provider`.

Reply:
Done. PTAL.